core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
-            
-            lease = decideWorkspace(node, Computer.currentComputer().getWorkspaceList());
+                
+                lease = decideWorkspace(node, Computer.currentComputer().getWorkspaceList());
 
-            workspace = lease.path.getRemote();
-            listener.getLogger().println(Messages.AbstractBuild_BuildingInWorkspace(workspace));
-            node.getFileSystemProvisioner().prepareWorkspace(AbstractBuild.this,lease.path,listener);
+                workspace = lease.path.getRemote();
+                listener.getLogger().println(Messages.AbstractBuild_BuildingInWorkspace(workspace));
+                node.getFileSystemProvisioner().prepareWorkspace(AbstractBuild.this,lease.path,listener);
 
-            for (WorkspaceListener wl : WorkspaceListener.all()) {
-                wl.beforeUse(AbstractBuild.this, lease.path, listener);
-            }
+                for (WorkspaceListener wl : WorkspaceListener.all()) {
+                    wl.beforeUse(AbstractBuild.this, lease.path, listener);
+                }
 
-            getProject().getScmCheckoutStrategy().preCheckout(AbstractBuild.this, launcher, this.listener);
-            getProject().getScmCheckoutStrategy().checkout(this);
+                getProject().getScmCheckoutStrategy().preCheckout(AbstractBuild.this, launcher, this.listener);
+                getProject().getScmCheckoutStrategy().checkout(this);
 
-            if (!preBuild(listener,project.getProperties()))
-                return Result.FAILURE;
+                if (!preBuild(listener,project.getProperties()))
+                    return Result.FAILURE;
 
-            Result result = doRun(listener);
+                result = doRun(listener);
+            } finally {
+                if (!tearDownMarker.tornDown) {
+                    // looks like environments are not torn down yet, do it now (in reverse order)
+                    boolean tearDownFailed = false;
+                    boolean tearDownInterrupted = false;
+                    for (int i = buildEnvironments.size() - 1; i >= 0; i--) {
+                        final Environment environment = buildEnvironments.get(i);
+                        try {
+                            if (!environment.tearDown(AbstractBuild.this, listener)) {
+                                tearDownFailed = true;
+                            }
+                        } catch (IOException | InterruptedException | RuntimeException e) {
+                            tearDownFailed = true;
+                            // exceptions are ignored to give a chance to all environments to tear down
+                            listener.error("Unable to tear down: " + e.getMessage());
+                            e.printStackTrace(listener.getLogger());
+                            if (e instanceof InterruptedException) {
+                                // don't forget we've been interrupted
+                                tearDownInterrupted = true;
```

Comment:
I'm not sure if this is the right approach since being interrupted usually means to exit ASAP

Reply:
I don't have a set opinion on what to do after the tear-down loop in case we've caught an `InterruptedException`. I've chosen to `Thread.currentThread().interrupt()`, but maybe we could as well raise a new `InterruptedException`, or maybe simply set an appropriate build status.

But I really think that completing the iteration on remaining environments, even in case of an `InterruptedException`, is the right thing to do. Leaving temporary files behind, or orphan processes, by not tearing down the remaining environments seems "wrong" to me (when it can be avoided). I think of multiple build wrappers (environments) not as a list (although that's how it is implemented), but rather as nested try/finally blocks.

Let's wait for more reviews before changing this code...

Reply:
> I don't have a set opinion on what to do after the tear-down loop in case we've caught an InterruptedException.

I've now changed my code to re-throw an `InterruptionException`. See https://github.com/jenkinsci/jenkins/pull/4517#discussion_r385416603 and 625b783.